### ISSUE TITLE
Safeguard functions of Autoscaler while MCM is offline

### ISSUE THREAD

**Issue by hardikdr** [gardener/autoscaler#60]
**What would you like to be added**:
We should investigate and identify the gaps in the behavior of autoscaler when MCM is offline. The following is a known impact that we know:
- Autoscaler may decide to scale-down the nodes, while MCM never acts on it as it's offline. Later restart of MCM can cause disruptions, and inconsistency. 

**Why is this needed**: For better robustness of both components Autoscaler and MCM, working together.

**Comment by vlerenc** [comment]
I am not really sure, what can be done here @hardikdr, if we want to keep both components separate. Do you have a proposal? It is clear, that we need to ensure that the control plane is always up - including MCM. Now, if MCM is down, you write, there are follow-up issues even if MCM gets back from the dead. Which complications, disruptions, and inconsistencies are these?

**Comment by amshuman-kr** [comment]
@hardikdr Thanks for filing the issue. As @vlerenc mentioned above, could you please elaborate what issues do you see if MCM is down but CA is up and then later MCM comes back up?

In principle, ideally CA should be able to tolerate MCM downtime gracefully because I would expect that to CA it would look no different from an unhealthy nodepool. There might be edge cases, of course. It would be good to know them.

**Comment by hardikdr** [comment]
Basically, the scale-down evaluation of autoscaler keeps watching the node-objects irrespective of MCM/autoscaling-group is active or not. Only by looking at the node-object, Autoscaler may decide to delete certain nodes(say 2), it would take the following steps scale-down:
1. Taint the 2 candidate node-objects with ~`DeletionCandidate`. [Though on the failure of removal, autoscaler removes these taints]
2. Find Machine objects related to the node-object, reduce their `priority` annotation to `1`.
3. Reduce the `.Spec.Replicas` of MachineDeployment by 2.


If MCM is running during these steps, it would simply delete those 2 machines along with machine-object, and nodes. 

But in case MCM is offline, `Spec.Replicas` of MachineDeployment is reduced by 2, but nothing really happens in the cloud. We still have those machines, registered as nodes, at some point capable of taking the new workload as CA removes the `DeletionCandidate` taint after a few failed-retries.
- If the new workload is introduced exactly during this time, it could get scheduled on the nodes which are candidates for the deletion. Also to mention, unless there are no `Pending` pods, autoscaler may not care to scale-up the replicas of MachineDeployment.

MCM, coming alive now, will simply try to respect it's `Spec.Replicas` and delete those 2 machines which might not be even deletion-candidate anymore, which is not favorable. 

A follow-up case, when there is a significant amount of workload introduced, can cause autoscaler to increase the `Spec.Replicas` of MachineDeployment to greater than the original. 
- In this case, we still have few machines which are annotated with lower priority.
- Ideally, we could revert the `priority` annotation immediately if scale-down fails, but it doesn't really fail from autoscaler PoV unless the timeout hit.

In summary, this is rather rare, and a short downtime of MCM should not cause an issue. But if the workload keeps changing during MCM is offline, could be problematic.


A simple resolution could be:
- MCM could expose a health-endpoint say `ready-for-autoscaling` via a k8s service, or simple `healthz` should also be fine.
- Autoscaler could then check the health of MCM by querying the service, before any scale-down/up, and also return an error during the [`Refresh()`](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L256) cloud-provider-call if MCM is not healthy.

**Comment by amshuman-kr** [comment]
Thanks for the explanation!

> A simple resolution could be:
> 
> * MCM could expose a health-endpoint say `ready-for-autoscaling` via a k8s service, or simple `healthz` should also be fine.
> * Autoscaler could then check the health of MCM by querying the service, before any scale-down/up, and also return an error during the [`Refresh()`](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L256) cloud-provider-call if MCM is not healthy.

I think the problem more general than can be addressed by this proposed solution. There might other reasons (e.g. issues in the infrastructure) that prevent a scale down from happening. I would guess the issues you mentioned could manifest in those cases as well.

How about enhancing CA to increment the `spec.replicas` whenever it removes the `DeletionCandidate` taint or priority annotation (preferably _just before_ removing the taint/annitation)?

**Comment by himanshu-kun** [comment]
## Post grooming decision
We could implement the [`Refresh`](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L328) call and return an error if the MCM is down after doing a health check, since if `Refresh` returns error then autoscaler doesn't do any operation, which is what we need.

**Comment by himanshu-kun** [comment]
~## More Unexpected behaviour cases (will keep adding more)~

~### Case 1~

~Three worker pools A, B, C~
~priority expander ,maxNodeProvisionTime=5 units of time,  priorities specified are:~
~C -> 30 (highest)~
~B -> 20~
~A -> 10~

~(unschedulable pods comes)~
~t=0 -> C scaled up (means md scaled by +1)~
~(Mcm down till t=6 , so no machine obj created)~
~t=5 -> C backoff (till t=10)~

~t=5 -> B scaled up~
~t=6 -> MCM launches machine in both B and C now (unexpected behavior)~

The above is wrong example pls ignore

**Comment by sssash18** [comment]
> Basically, the scale-down evaluation of autoscaler keeps watching the node-objects irrespective of MCM/autoscaling-group is active or not. Only by looking at the node-object, Autoscaler may decide to delete certain nodes(say 2), it would take the following steps scale-down:
> 
> 1. Taint the 2 candidate node-objects with ~`DeletionCandidate`. [Though on the failure of removal, autoscaler removes these taints]
> 2. Find Machine objects related to the node-object, reduce their `priority` annotation to `1`.
> 3. Reduce the `.Spec.Replicas` of MachineDeployment by 2.
> 
> If MCM is running during these steps, it would simply delete those 2 machines along with machine-object, and nodes.
> 
> But in case MCM is offline, `Spec.Replicas` of MachineDeployment is reduced by 2, but nothing really happens in the cloud. We still have those machines, registered as nodes, at some point capable of taking the new workload as CA removes the `DeletionCandidate` taint after a few failed-retries.
> 
> * If the new workload is introduced exactly during this time, it could get scheduled on the nodes which are candidates for the deletion. Also to mention, unless there are no `Pending` pods, autoscaler may not care to scale-up the replicas of MachineDeployment.
> 
> MCM, coming alive now, will simply try to respect it's `Spec.Replicas` and delete those 2 machines which might not be even deletion-candidate anymore, which is not favorable.
> 
> A follow-up case, when there is a significant amount of workload introduced, can cause autoscaler to increase the `Spec.Replicas` of MachineDeployment to greater than the original.
> 
> * In this case, we still have few machines which are annotated with lower priority.
> * Ideally, we could revert the `priority` annotation immediately if scale-down fails, but it doesn't really fail from autoscaler PoV unless the timeout hit.
> 
> In summary, this is rather rare, and a short downtime of MCM should not cause an issue. But if the workload keeps changing during MCM is offline, could be problematic.
> 
> A simple resolution could be:
> 
> * MCM could expose a health-endpoint say `ready-for-autoscaling` via a k8s service, or simple `healthz` should also be fine.
> * Autoscaler could then check the health of MCM by querying the service, before any scale-down/up, and also return an error during the [`Refresh()`](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L256) cloud-provider-call if MCM is not healthy.

We actually reproduced the scenarios and following were the outputs:
1.  In case of scale down when the mcm is offline the taints from the nodes is never removed by the autoscaler and those nodes are deleted whenever mcm comes back.
2. In case of scale up when the mcm is offline the tainted nodes are considered by the autoscaler but the scheduler never schedule pods on it due to the taint, so new nodes comes up as soon as mcm is online.
